src/utils/contractCaller.ts

```
@@ -0,0 +1,56 @@
+import { ErrorDescription, ethers } from "ethers";
+import { errorAbis } from "./ContractErrors";
+import { Bytes } from "@/types/types";
+
+// interface for parsing errors
+const errorInterface = new ethers.Interface(errorAbis);
+
+export interface CustomError {
+    name: string;
+    args: any[];
+    signature: string;
+}
+
+/**
+ * Custom error class that extends Error and includes decoded Solidity error information
+ */
+export class CustomContractError extends Error {
+    public readonly errorDescription: ErrorDescription;
+    public readonly isCustomError = true;
+    public readonly originalError: any;
+
+    constructor(customError: ErrorDescription, originalError: any) {
+        super(
+            `${customError.name}${customError.args.length > 0 ? ` (args: ${customError.args.join(", ")})` : ""}`
+        );
+        this.name = customError.name;
+        this.errorDescription = customError;
+        this.originalError = originalError;
+    }
+}
+
+function decodeCustomError(errorData: Bytes): ErrorDescription | null {
+    if (!errorData || errorData.length < 10) return null;
+
+    return errorInterface.parseError(errorData);
+}
+
+export function isCustomContractError(
+    error: any
+): error is CustomContractError {
+    return !!error && error.isCustomError === true;
+}
+
+export async function call<T>(contractCall: () => Promise<T>): Promise<T> {
+    try {
+        return await contractCall();
+    } catch (error: any) {
+        const customError = error.data ? decodeCustomError(error.data) : null;
+
+        if (customError) {
+            throw new CustomContractError(customError, error);
+        }
+
+        throw error;
+    }
```

I don't like this call function - Imo we shouldn't change the normal usage or flow of interacting with the contract and if we do better to do with `proxies` - maybe I'm biased towards proxies :D 

I like the idea of having a `CustomEvmError` class that we use internally within the SDK and throw that instead of what ethers throws - so we can inject that transformation with proxies and/or just transform `error.data` into our `CustomEvmError` when interacting with `EVM.call` directly -> and then we just interpret our own `CustomEvmError` that's already decoded

So we don't break any logical flow - we use everything the same internally. If we want to catch the error we catch it and do something (only difference it's now always decoded), else let it throw and log its description 

Imo all of this code here and the code from `ContractErrors` should be in one place `CustomEvmError` and we don't need the ContractError.enum - we'll just interpret our decoded custom error

---

not a very good ideas i think
with proxy we will lose the type safety and autocomplete when interacting with the proxied contract
this will drag us down a hole of rebuilding the method surface of the contract OR not having autocomplete

i think you are indeed a bit biased towards proxies

---

Why do you think we lose type safety or autocomplete? Everything stays the same - take a look at debug proxy 

The only thing that the proxy does is catch the ethers error -> decode it and throw our error 

Everything else stays the same 

Only if we do a try catch we expect our error instead of ethers serialized error

---

ill give it a shot. see how it goes